ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosChangeFeedBasedReplicationFeed.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    return new CosmosChangeFeedFindToken(cosmosChangeFeedFindToken.getBytesRead() + resultSize,
+        changeFeedCache.get(partitionId).getStartContinuationToken(),
+        changeFeedCache.get(partitionId).getEndContinuationToken(), index,
+        changeFeedCache.get(partitionId).getFetchedEntries().size(),
+        changeFeedCache.get(partitionId).getAzureRequestId(), cosmosChangeFeedFindToken.getVersion());
+  }
+
+  @Override
+  public CosmosChangeFeedFindToken getNextEntriesAndUpdatedToken(FindToken curfindToken,
+      List<CloudBlobMetadata> nextEntries, long maxTotalSizeOfEntries, String partitionPath)
+      throws DocumentClientException {
+    return getNextEntriesAndToken((CosmosChangeFeedFindToken) curfindToken, nextEntries, maxTotalSizeOfEntries,
+        partitionPath);
+  }
+
+  /**
+   * Check is the cache is valid for the {@code cosmosChangeFeedFindToken} provided.
+   * @param partitionId partition of the {@code cosmosChangeFeedFindToken}.
+   * @param cosmosChangeFeedFindToken {@link CosmosChangeFeedFindToken} object.
+   * @return true is cache is valid. false otherwise.
+   */
+  private boolean isCacheValid(String partitionId, CosmosChangeFeedFindToken cosmosChangeFeedFindToken) {
+    ChangeFeedCacheEntry changeFeedCacheEntry = changeFeedCache.get(partitionId);
+    return Utils.checkNullableStringEquals(cosmosChangeFeedFindToken.getAzureTokenRequestId(),
+        changeFeedCacheEntry.getAzureRequestId()) && Utils.checkNullableStringEquals(
+        cosmosChangeFeedFindToken.getStartContinuationToken(), changeFeedCacheEntry.getStartContinuationToken())
+        && Utils.checkNullableStringEquals(cosmosChangeFeedFindToken.getEndContinuationToken(),
+        changeFeedCacheEntry.getEndContinuationToken())
+        && cosmosChangeFeedFindToken.getTotalItems() == changeFeedCacheEntry.getFetchedEntries().size();
+  }
+
+  private boolean cacheEmpty(String partitionId) {
+    return changeFeedCache.get(partitionId).getFetchedEntries().size() == 0;
+  }
+
+  /**
+   * Populate change feed cache by querying cosmos for the next set of change feed entries after the specified request continuation token.
+   * @param partitionId Partition for which the change feed cache needs to be populated.
+   * @param startRequestContinuationToken request continuation token from which the change feed query needs to be made.
+   */
+  private void populateChangeFeedCache(String partitionId, String startRequestContinuationToken)
+      throws DocumentClientException {
+    List<CloudBlobMetadata> changeFeedEntries = new ArrayList<>(defaultCacheSize);
+    String newRequestContinuationToken =
+        cosmosDataAccessor.queryChangeFeed(startRequestContinuationToken, defaultCacheSize, changeFeedEntries,
+            partitionId);
+    ChangeFeedCacheEntry changeFeedCacheEntry =
+        new ChangeFeedCacheEntry(startRequestContinuationToken, newRequestContinuationToken,
+            UUID.randomUUID().toString(), changeFeedEntries);
```

Comment:
We shouldn't need to generate a new requestid on every call.  Can reuse the previous one if not first time.